[-Wunsafe-buffer-usage][BoundsSafety] support __null_terminated in the interop mode

ziqingluo-90 opened the pull request:
For a __null_terminated pointer 'p', only a __null_terminated pointer, a string literal or a call to `std::string::c_str` can be assigned to (/passed to/used to initialize) 'p'.  Otherwise, in the interop mode, an -Wunsafe-buffer-usage warning will be reported.

Similar to how __counted_by is handled, incompatibility in nested __null_terminated pointers (e.g., input/output pointer type arguments) is reported as Bounds Safety type checking errors.

(rdar://138798346)

ziqingluo-90 committed:
[-Wunsafe-buffer-usage][BoundsSafety] support __null_terminated in the interop mode

For a __null_terminated pointer 'p', only a __null_terminated pointer,
a string literal or a call to `std::string::c_str` can be assigned to
(/passed to/used to initialize) 'p'.  Otherwise, in the interop mode,
an -Wunsafe-buffer-usage warning will be reported.

Similar to how __counted_by is handled, incompatibility in nested
__null_terminated pointers (e.g., input/output pointer type arguments)
is reported as Bounds Safety type checking errors.

(rdar://138798346)

ziqingluo-90 commented:
@swift-ci test

ziqingluo-90 commented:
CC: @dtarditi

ziqingluo-90 commented:
@swift-ci test

ziqingluo-90 commented:
@swift-ci test

dtarditi commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12186,13 +12186,39 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
+        if (LangOpts.CPlusPlus && SafeBuffersEnabled) {
+          if (const auto *MCE =
+                  dyn_cast<CXXMemberCallExpr>(RHSExpr->IgnoreParenImpCasts())) {
+            IdentifierInfo *MethodDeclII = nullptr, *ObjTypeII = nullptr;
```
I think the variable names could be better here.   MethodDeclI1 and ObjTypeI1 are hard to read.

ziqingluo-90 replied:
They are not `MethodDeclI1` or `ObjTypeI1`.  They are two capital "i"s.  I saw some other places in the code base where variables holding `IdentifierInfo` are named `II`, so I just followed the convention.

dtarditi commented on clang/test/SemaCXX/warn-unsafe-buffer-usage-libc-functions.cpp:
```diff
@@ -155,3 +158,47 @@ void ff(char * p, char * q, std::span<char> s, std::span<char> s2) {
+
+#include <ptrcheck.h>
+typedef unsigned size_t;
+
```
Some of these new tests look unrelated to this change.  Please consider splitting them out to a separate PR.

ziqingluo-90 replied:
sure.  These are things left un-upstreamed.

dtarditi commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12186,13 +12186,39 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
+              if (RecordDecl->isInStdNamespace())
+                ObjTypeII = RecordDecl->getIdentifier();
+            }
+            if (MethodDeclII && ObjTypeII &&
```
I think this could be combined with the prior if-statement for checking that RecordDecl is in the standard namespace. It would be clearer.

dtarditi commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -13603,6 +13603,19 @@ def note_unsafe_count_attributed_pointer_argument : Note<
+                  "initialization of a pointer of|"
+                  "assignment to a pointer of|"
+                  "casting to}2 '__null_terminated' type; "
+  "only '__null_terminated' pointers, string literals, and 'std::string::c_str' calls are compatible with '__null_terminated' pointers">,
```
It might be simpler to say that. the source pointer may not point to null-terminated data.  I could imagine that we eventually we will increase the number of cases that we allow.  I don't have super strong opinions about this - it is nice to tell people what is allowed, but it doesn't tell them what the error is.

ziqingluo-90 replied:
I see. Yeah, I should explain what the error is.   The convention for warning messages (at least for -Wunsafe-buffer-usage) is that we have a short but clear enough sentence to just say "something is unsafe", then put necessary explanation in notes.   My message here might already be lengthy.  I will explain what the error is in a note and also move  `"only '__null_terminated' pointers, ..."` (sort of the suggestion) to a note too.

ziqingluo-90 replied:
oh, I maybe misunderstood part of your comment here.  
You were suggesting to simply saying "source pointer may not point to null-terminated data." instead of breaking into cases.    I'm not against the idea of making the warning message simpler.  My only concern is that warning messages under `-Wunsafe-buffer-usage` all have the forms "unsafe [some operation]" or "[some operation] is unsafe".  I'd like this message to follow that.

Different cases are determined by `AssignmentAction`, which is handled by the existing framework in `Sema`.  I assume it is stable.

ziqingluo-90 replied:
Or maybe `"unsafe conversion to __null_terminated type; source pointer may not point to null-terminated data."` ?  If "conversion" is a appropriate umbrella term for these operations.

dtarditi reviewed (commented):
I have some suggestions.

hnrklssn commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -13603,6 +13603,19 @@ def note_unsafe_count_attributed_pointer_argument : Note<
+  "only '__null_terminated' pointers, string literals, and 'std::string::c_str' calls are compatible with '__null_terminated' pointers">,
+  InGroup<UnsafeBufferUsage>, DefaultIgnore;
+def warn_unsafe_named_cast_to_null_terminated_pointer : Warning<
+  "C++ named cast to '__null_terminated' type is unsafe">,
```
This may be overly specific: any cast to `__null_terminated` is unsafe. If I got this error message I'd try doing a C-style cast instead, which would not help.

hnrklssn replied:
Wait, looking at the test cases it seems like C-style casts are deliberately allowed. What's the rationale?

ziqingluo-90 replied:
My original idea is that it should be rare for C++ programmers to cast null_terminated pointers with named casts.  And, Bounds Safety's type checking for null_terminated can only be re-used for C-Style casting. So why not force C++ programmers to just use C-Style casts for null_terminated pointers?

But now I changed my mind.  I think it is too strict.  Moreover, it will be difficult to relax the restriction once this warning is added.  On the other hand, we are allowed to have some false negatives for some rare use cases and make warnings for them later.   So I think I will remove this warning.   @rapidsna @patrykstefanski  WDYT?

Thanks for bringing this up!

hnrklssn replied:
I disagree that it's easier to add warnings later than relaxing later, since relaxing the rules doesn't disrupt existing adopters while new warnings do. But that said, if there's no reason for preferring one cast over the other it seems reasonable to remove it.

rapidsna replied:
Where's the test case to allow C-Style casts?

rapidsna commented on clang/test/SemaCXX/warn-unsafe-buffer-usage-libc-functions.cpp:
```diff
@@ -4,7 +4,10 @@
+// RUN: %clang_cc1 -std=c++20 -Wno-all -Wunsafe-buffer-usage \
+// RUN:            -verify -fexperimental-bounds-safety-attributes %s -DINTEROP_MODE
 
+#ifndef INTEROP_MODE
```
Looks like the code isn't actually shared for INTEROP_MODE vs !INTEROP_MODE. Shouldn't it be just a separate test file?

ziqingluo-90 replied:
sure. These are things left un-upstreamed.

rapidsna commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -238,6 +238,17 @@ namespace {
               DstPointerType->getTerminatorValue(Self.Context).isZero();
+          if (Self.LangOpts.isBoundsSafetyAttributeOnlyMode() &&
+              Self.LangOpts.CPlusPlus &&
+              !Self.Diags.isIgnored(
```
Why is it controlled by `diag::warn_unsafe_assign_to_null_terminated_pointer`, not by the UnsafeBufferUsage?

ziqingluo-90 replied:
Usually, we use `diag::warn_unsafe_buffer_operation` to indicate whether `-Wunsafe-buffer-usage` is on for a source location. [(For example: here.)](https://github.com/llvm/llvm-project/blob/77a8338fa65b959794c4ab9b2afd1d08deed49fe/clang/lib/Sema/AnalysisBasedWarnings.cpp#L2570C32-L2570C60)

The `diag::warn_unsafe_assign_to_null_terminated_pointer` is under the UnsafeBufferUsage warning group and is only controlled by the `-Wunsafe-buffer-usage` flag.     So functionally it is equivalent to using `diag::warn_unsafe_buffer_operation`.  I originally thought it is more specific but now I feel like using `diag::warn_unsafe_buffer_operation` is more consistent and probably more readable.   I will change it to `diag::warn_unsafe_buffer_operation`, if this is your concern.

rapidsna replied:
Yes, consistently using `diag::warn_unsafe_buffer_operation` sounds a more robust implementation rather than relying on different warnings for each individual set of warnings. 

Nit: You might also want to have a helper function like `IsSafeBufferMode` and use that instead to avoid mistakes and duplication.

rapidsna commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -238,6 +238,17 @@ namespace {
+            // In C++ Safe Buffers/Bounds Safety interop mode, the compiler
+            // reports a warning under -Wunsafe-buffer-usage:
+            DiagKind = diag::warn_unsafe_assign_to_null_terminated_pointer;
+            isInvalid = true;
```
Shouldn't it be `isInvalid = false` because this is a warning. Then, it's controlled by -Werror if the compiler still wants to fail the build.

ziqingluo-90 replied:
yes, it is a mistake!

rapidsna commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -576,6 +587,17 @@ Sema::BuildCXXNamedCast(SourceLocation OpLoc, tok::TokenKind Kind,
+    if (DestType->getAs<ValueTerminatedType>()) {
+      Diag(E->getBeginLoc(),
+           diag::warn_unsafe_named_cast_to_null_terminated_pointer);
+      return ExprError();
```
Similarly, it's weird that it returns `ExprError()` on a warning. It should either an error or just pass through.

ziqingluo-90 replied:
I will remove it.

rapidsna commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -576,6 +587,17 @@ Sema::BuildCXXNamedCast(SourceLocation OpLoc, tok::TokenKind Kind,
 
+  // In C++ Safe Buffers/Bounds Safety interop mode, forbidding  for now all
+  // named cast to __null_terminated pointer type.
+  if (DestType->isPointerType() && LangOpts.hasBoundsSafetyAttributes() &&
```
Did you mean `Self.LangOpts.isBoundsSafetyAttributeOnlyMode()`?

ziqingluo-90 replied:
This is handling C++ constructs interacting with bounds attributes.  So my original thinking is that the guard is whenever bounds attributes is visible to C++ code.

ziqingluo-90 replied:
My general idea is that the interoperation work should be guarded by `LangOpts.hasBoundsSafetyAttributes` combined with the testing of whether`-Wunsafe-buffer-usage` is enabled for most places.  We only use `LangOpts.isBoundsSafetyAttributeOnlyMode()` to guard against the work that otherwise can affect `-fbounds-safety`.  Let me know if this is not the best way of thinking.

rapidsna replied:
`LangOpts.hasBoundsSafetyAttributes` is enabled in both the annotation only mode and the full firebloom mode. I was asking because you may want an error and may be a different diagnostics in the firebloom mode, which we didn't design yet. The current way of doing it is still fine for this patch and we can adjust it later for C++ firebloom mode.

rapidsna commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -576,6 +587,17 @@ Sema::BuildCXXNamedCast(SourceLocation OpLoc, tok::TokenKind Kind,
+  // In C++ Safe Buffers/Bounds Safety interop mode, forbidding  for now all
+  // named cast to __null_terminated pointer type.
+  if (DestType->isPointerType() && LangOpts.hasBoundsSafetyAttributes() &&
+      !Diags.isIgnored(diag::warn_unsafe_named_cast_to_null_terminated_pointer,
```
Again, why it's guarded by this specific diag, not the unsafe buffer usage flag?

rapidsna commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12186,13 +12186,39 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
         }
+        // If in C++ Safe Buffers/Bounds Safety interoperation mode, the RHS can
+        // be 'std::string::c_str':
+        bool SafeBuffersEnabled = !Diags.isIgnored(
```
Why `SafeBuffersEnabled` is determined by `diag::warn_unsafe_assign_to_null_terminated_pointer`? Shouldn't it be the combination of the flag for UnsafeBufferUsage and `getLangOpts().isBoundsSafetyAttributeOnlyMode()`?

ziqingluo-90 replied:
No.  We are in `CheckValueTerminatedAssignmentConstraints` function.  So it must already checked if `LangOpts.hasBoundsSafetyAttributes()`.  We just need to check if `-Wunsafe-buffer-usage` is enabled here.  As I explained above, I will change `diag::warn_unsafe_assign_to_null_terminated_pointer` to   `diag::warn_unsafe_buffer_operation`.

rapidsna commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -13603,6 +13603,19 @@ def note_unsafe_count_attributed_pointer_argument : Note<
 def warn_unsafe_single_pointer_argument : Warning<
   "unsafe assignment to function parameter of __single pointer type">,
   InGroup<UnsafeBufferUsage>, DefaultIgnore;
+def warn_unsafe_assign_to_null_terminated_pointer : Warning<
```
This is probably almost a duplicate of an existing one `warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by`. I guess you introduced a new one because you want something in the `UsafeBufferUsage` group? Okay, I see the diagnostic message at the end is different.

ziqingluo-90 replied:
`warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by` mentions "indexable" and "unsafe_forge", which do not exist in the interop mode.  This is the main reason for having another diagnostic message.

patrykstefanski replied:
You could move those "indexable" and "unsafe_forge" to `%select` in `warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by` and not print them in interop.

ziqingluo-90 committed:
Remove changes in warn-unsafe-buffer-usage-libc-functions.cpp because of irrelevance.

ziqingluo-90 committed:
Address comments

- Make the diagnostic message simply saying "unsafe [conversion
  operation] of '__null_terminated' type"; further explanation is put
  in notes.
- Use `warn_unsafe_buffer_operation` to consistently test if CXX Safe
  Buffer is enabled.  Add an helper function for the test.
- Remove the warning on using C++ named cast to obtain
  __null_terminated pointers.
- Added tests for good cases of C-Style casts.
- Consistently use isBoundsSafetyAttributeOnlyMode for code involving
  bounds attributes but not shared with full Bounds Safety mode.

patrykstefanski commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -205,7 +205,7 @@ namespace {
       /// int *__indexable p2 = (int *__indexable)p1;
       /// \endcode
-      if (Self.getLangOpts().BoundsSafety) {
+      if (Self.getLangOpts().hasBoundsSafetyAttributes()) {
```
Nit: We want to remove `hasBoundsSafetyAttributes()`, since it is not needed anymore. `Self.getLangOpts().hasBoundsSafetyAttributes()` -> `Self.getLangOpts().BoundsSafetyAttributes`.

patrykstefanski commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1560,6 +1560,11 @@ class Sema final : public SemaBase {
   BoundsSafetyPointerAttributes CurPointerAbi;
+
+  /// \returns true iff `-Wunsafe-buffer-usage` is enabled for `Loc`.
+  bool isCXXSafeBuffersEnabledAt(SourceLocation Loc) {
```
Nit: Can this method be `const`?

patrykstefanski commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12186,13 +12186,35 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
+              // match fails:
+              ObjTypeII = RecordDecl->getIdentifier();
+            if (MethodDeclII && ObjTypeII &&
+                MethodDeclII->getName() == "c_str" &&
```
If there a reason we support `std::string::c_str()` but not `std::string::data()`?

ziqingluo-90 replied:
`std::string::data()` is not null-termination guaranteed as I remember.

patrykstefanski replied:
https://en.cppreference.com/w/cpp/string/basic_string/c_str
`c_str() and data() perform the same function. (since C++11)`

ziqingluo-90 replied:
Aha, it's changed since C++11.

patrykstefanski commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12186,13 +12186,35 @@ Sema::CheckValueTerminatedAssignmentConstraints(QualType LHSType,
+            if (MethodDeclII && ObjTypeII &&
+                MethodDeclII->getName() == "c_str" &&
+                ObjTypeII->getName() == "basic_string")
+              return Sema::Compatible;
```
Does this work if LHS is terminated by a value different than 0, e.g. `const char *__terminated_by(42)`?

ziqingluo-90 replied:
Good point.   Need to check it is 0-terminator.

patrykstefanski commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -238,6 +238,15 @@ namespace {
+            DiagKind = diag::warn_unsafe_assign_to_null_terminated_pointer;
+            isInvalid = false;
+            break;
+          }
```
Same here, we don't check if the terminator is 0.

ziqingluo-90 replied:
I have changed it to re-use `warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by`.  We do not need to check terminator there now.

ziqingluo-90 committed:
Address more comments.

- Change LangOpts.hasBoundsSafetyAttributes() to LangOpts.BoundsSafetyAttributes
- Correct that std::string::c_str() should be not compatible with (non-0)-terminated pointers

ziqingluo-90 committed:
Remove warn_unsafe_assign_to_null_terminated_pointer

Reuse the existing warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by

ziqingluo-90 committed:
Add that std::string::data() is also null-terminated

ziqingluo-90 commented:
I have addressed comments by:
- Remove irrelevant changes in `test/SemaCXX/warn-unsafe-buffer-usage-libc-functions.cpp`
- Reuse `warn_bounds_safety_incompatible_non_terminated_by_to_terminated_by` for interop's diagnostic.  Hide irrelevant texts in "%select".
- Use `warn_unsafe_buffer_operation` to consistently test if CXX Safe Buffer is enabled.
- Remove the warning on __null_terminated casting with C++ named casts.
- Add tests for OK cases of C-Style casts.
- Consistently use `isBoundsSafetyAttributeOnlyMode` to guard the code that involves bounds attributes but is not shared with full Bounds Safety mode.
- Change LangOpts.hasBoundsSafetyAttributes() to LangOpts.BoundsSafetyAttributes
- `std::string::c_str/data()` should only be compatible with 0-terminated pointers

@patrykstefanski @rapidsna @hnrklssn @dtarditi  please take another look

ziqingluo-90 commented:
@swift-ci test

rapidsna commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -20694,6 +20720,9 @@ bool Sema::DiagnoseAssignmentResult(AssignConvertType ConvTy,
       AssignedVarFixIts;
   VarDecl *BoundsSafetyIncompatNestedPtrLocalVarToFix = nullptr;
   FixItHint BoundsSafetyIncompatNestedPtrFixIt;
+  bool InCXXBoundsSafetyMode = LangOpts.CPlusPlus &&
```
Nit: This seems to be duplicated too in another Sema file. So it's another candidate to get a helper function that tales `Loc`.

ziqingluo-90 replied:
👌

hnrklssn commented on clang/lib/Sema/SemaCast.cpp:
```diff
@@ -200,12 +200,17 @@ namespace {
+      if (Self.getLangOpts().BoundsSafetyAttributes) {
         unsigned DiagKind = 0;
         bool isInvalid = false;
         // The type error may be nested, so any pointer can result in VT errors
```
Do we want to assert that `Self.getLangOpts().BoundsSafety || ConvertTy == Sema::Compatible || (ConvertTy == Sema::IncompatibleNonValueTerminatedToValueTerminatedPointer && InCXXBoundsSafetyMode)`? Especially since `CheckValueTerminatedAssignmentConstraints` used to only run under `BoundsSafety`, but now it runs under `BoundsSafetyAttributes` even for C and in C++ without safe buffers mode. We really don't want it to accidentally start returning errors in these modes. Otherwise LGTM

ziqingluo-90 replied:
Fair point.  But maybe the following is enough?
`Self.getLangOpts().BoundsSafety ||  Self.isCXXSafeBuffersBoundsSafetyInteropEnabledAt(SrcExpr.get()->getBeginLoc())` ,
where `isCXXSafeBuffersBoundsSafetyInteropEnabledAt` is defined by
```
  bool isCXXSafeBuffersBoundsSafetyInteropEnabledAt(SourceLocation Loc) const {
    return LangOpts.CPlusPlus && LangOpts.isBoundsSafetyAttributeOnlyMode() &&
           isCXXSafeBuffersEnabledAt(Loc);
  }
```

hnrklssn replied:
That works for me 👍

ziqingluo-90 committed:
Address minor comments

- Refactor interop mode check to a helper function
- Guard the type checking for C-style cast so that it runs only in full bounds safety or interop only mode

ziqingluo-90 committed:
Use `BoundsSafety || isCXXSafeBuffersBoundsSafetyInteropEnabledAt(...)` in SemaInit

ziqingluo-90 commented:
I plan to merge this PR this weekend.  @hnrklssn @rapidsna
